Misc/Adam.cpp

```diff
@@ -2673,47 +2674,176 @@ int IsGridDotted(COMMAND_T* = NULL)
 	return r < 0.000001 || r > 0.66666;
 }
 
+int IsGridSwing(COMMAND_T* = NULL)
+{
+	return GetToggleCommandStateEx(0, 42304); // Grid: Toggle swing grid
+}
+
+int IsAWSetGridPreserveType(COMMAND_T* ct)
+{
+	int toggleState = 0;
+
+	double normalizedGrid = 0;
+	if ((int)ct->user < 0)
+		normalizedGrid = static_cast<double>(abs((int)ct->user)) * 4.0;
+	else
+		normalizedGrid = 4.0 / static_cast<double>((int)ct->user);
+
+	ConfigVar<double> pGridDiv("projgriddiv");
+	if (IsGridTriplet())
+	{
+		if (*pGridDiv == normalizedGrid * (2.0 / 3.0))
+			toggleState = 1;
+	}
+	else if (IsGridDotted())
+	{
+		if (*pGridDiv == normalizedGrid * (3.0 / 2.0))
+			toggleState = 1;
+	}
+	else
+	{
+		if (*pGridDiv == normalizedGrid)
+			toggleState = 1;
+	}
+
+	return toggleState;
+}
 
 void AWToggleDotted(COMMAND_T* = NULL);
+void AWToggleSwing(COMMAND_T* = NULL);
 
 void AWToggleTriplet(COMMAND_T* = NULL)
 {
 	ConfigVar<double> pGridDiv("projgriddiv");
 
 	if (IsGridDotted())
 		AWToggleDotted();
+	if (IsGridSwing())
+		AWToggleSwing();
 
 	if (IsGridTriplet())
 		*pGridDiv *= 3.0/2.0;
 	else
 		*pGridDiv *= 2.0/3.0;
 
-
+	if (MIDIEditor_GetActive() && GetToggleCommandStateEx(SECTION_MIDI_EDITOR, 41022))     // Grid: Use the same grid division in arrange view and MIDI editor
+		if (IsGridTriplet() && !GetToggleCommandStateEx(SECTION_MIDI_EDITOR, 41004))       // Grid: Set grid type to triplet
+			MIDIEditor_LastFocused_OnCommand(41004, false);                                // Grid: Set grid type to triplet
+		else if (!IsGridTriplet() && !GetToggleCommandStateEx(SECTION_MIDI_EDITOR, 41003)) // Grid: Set grid type to straight
```

Comment:
While this code does behave as the indentation leads to believe, braces should be added around the first `if` to fix these warnings from Clang:

```
../Misc/Adam.cpp:2732:3: warning: add explicit braces to avoid dangling else [-Wdangling-else]
                else if (!IsGridTriplet() && !GetToggleCommandStateEx(SECTION_MIDI_EDITOR, 41003)) // Grid: Set grid type to straight
                ^
../Misc/Adam.cpp:2755:3: warning: add explicit braces to avoid dangling else [-Wdangling-else]
                else if (!IsGridDotted() && !GetToggleCommandStateEx(SECTION_MIDI_EDITOR, 41003)) // Grid: Set grid type to straight
                ^
2 warnings generated.
```

Reply:
So add braces then :)

Reply:
Btw, I like to fix warnings too ;) But I don't remember seeing that one!
But for sake of readability, I always skip braces when I can...it's just my style

Reply:
You can find Clang's warnings in the Travis build log for this PR: https://travis-ci.com/reaper-oss/sws/jobs/268161882#L460.

I'd prefer it if you'd amend the commit and force-pushed the branch yourself. I can do it if you want me to (as long as "Allow edits from maintainers" is enabled to the right of this page), but your local branch will diverge.

Reply:
I just look at VS warnings, but it had some many warnings I didn't catch this one.

Reply:
I'll enable it

Reply:
It was already enabled ;D